vdr/didx509/resolver.go

```diff
@@ -0,0 +1,222 @@
+func findValidationCertificate(metadata *resolver.ResolveMetadata, chain []*x509.Certificate) (*x509.Certificate, error) {
+	var validationCert *x509.Certificate
+	var err error
+	hashHeader, found := metadata.GetProtectedHeaderString(X509CertThumbprintHeader)
+	if found {
+		validationCert, err = findCertificateByHash(chain, hashHeader, HashSha1)
+		if err != nil {
+			return nil, err
+		}
+	}
+	hash256Header, found := metadata.GetProtectedHeaderString(X509CertThumbprintS256Header)
+	if found {
+		otherValidationCert, err := findCertificateByHash(chain, hash256Header, HashSha256)
+		if err != nil {
+			return nil, err
+		}
+		if validationCert == nil {
+			validationCert = otherValidationCert
+		} else {
+			if !otherValidationCert.Equal(validationCert) {
+				return nil, ErrNoMatchingHeaderCredentials
+			}
+		}
+	}
+	if validationCert == nil {
+		return nil, ErrNoCertsInHeaders
+	}
+	return validationCert, nil
+}
+
+// createDidDocument generates a new DID Document based on the provided DID identifier and validation certificate.
+func createDidDocument(id did.DID, validationCert *x509.Certificate) (*did.Document, error) {
+	didUrl := did.DIDURL{DID: id, Fragment: "0"}
+	verificationMethod, err := did.NewVerificationMethod(didUrl, ssi.JsonWebKey2020, id, validationCert.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	document := &did.Document{
+		Context: []interface{}{
+			ssi.MustParseURI("https://www.w3.org/ns/did/v1"),
+		},
+		ID:                 id,
+		Controller:         []did.DID{id},
+		VerificationMethod: did.VerificationMethods{verificationMethod},
+	}
+	document.AddKeyAgreement(verificationMethod)
+	document.AddAssertionMethod(verificationMethod)
+	document.AddAuthenticationMethod(verificationMethod)
+	document.AddCapabilityDelegation(verificationMethod)
+	document.AddCapabilityInvocation(verificationMethod)
+	return document, nil
+}
+
+// parseX509Did parses a DID (Decentralized Identifier) in the x509 format and returns a corresponding X509DidReference.
+func parseX509Did(id did.DID) (*X509DidReference, error) {
+	ref := X509DidReference{}
+	fullDidString := id.ID
+	policyStrings := []string{}
+	didString := ""
+	fullDidParts := strings.Split(fullDidString, "::")
+	if len(fullDidParts) == 1 {
+		didString = fullDidParts[0]
+	} else if len(fullDidParts) > 1 {
+		didString = fullDidParts[0]
+		policyStrings = fullDidParts[1:]
+	}
+	didParts := strings.Split(didString, ":")
+	if len(didParts) != 3 {
+		return nil, ErrDidMalformed
+	}
+
+	if didParts[0] != "0" {
+		return nil, ErrDidVersion
+	}
+	ref.Method = HashAlgorithm(didParts[1])
+	ref.RootCertRef = didParts[2]
+
+	for _, policyString := range policyStrings {
+		policyFragments := strings.Split(policyString, ":")
+		if len(policyFragments) > 1 {
+			policy := X509DidPolicy{Name: PolicyName(policyFragments[0]), Value: strings.Join(policyFragments[1:], ":")}
```

Comment:
Does this work with percent-encoded values in the `did:x509`? The spec (https://trustoverip.github.io/tswg-did-x509-method-specification/#san-policy) states:

> ```
policy-name     = "san"
policy-value    = san-type ":" san-value
san-type        = "email" / "dns" / "uri"
san-value       = 1*idchar
```
> san-type is the SAN type and must be one of email, dns, or uri. Note that dn is not supported.
> san-value is the SAN value, percent-encoded.